### ISSUE TITLE
TEST Llama against the original Llama implementation

### ISSUE THREAD

**Issue by PABannier** [ggerganov/llama.cpp#2630]
Hello Llama.cpp team!

I am currently working on [bark.cpp](https://github.com/PABannier/bark.cpp), but I'm struggling to have results that are close to the original implementation of [bark](https://github.com/suno-ai/bark). 

`bark` is made of 3 GPT encoders. I would like to check that the forward pass of these GPT models yields the same output as the original GPT models in Python. For now, I'm writing unit tests where I'm feeding the same token sequence to the Python and C++ model and I'm testing the logits.

I would expect to have small errors around 0.001 between the original logits and the logits output by `bark.cpp`. However, for the three encoders I end up having errors one or two orders of magnitude higher (0.05 to 0.1) than my 0.001 threshold. This seems quite high and above the typical margins of error you can expect from numerical errors.

I would like to know if some of you have tested the logits yielded by `llama.cpp` against the original Llama ones. Do you have an idea of what an acceptable error threshold should be? Besides, do you have any guidance regarding how to test the model forward pass?

Currently, I'm testing the logits and the sequence generated by the GPT by sampling deterministically (argmax sampling) to remove any stochasticity in my test. 

Thanks in advance

Pinging @ggerganov as he might have implemented tests after his first implementation of llama.cpp

**Comment by ggerganov** [comment]
I haven't tested LLaMA, but I did tests when implementing the GPT examples and SAM. The differences between the C++ and Python models a very small - well within numerical error ranges.

`ggml` code is difficult to debug. When I compare against a reference implementation, I usually do the following:

- start at the beginning of the computation graph and set the name of the first result to `check`:
  ```c++
  ggml_set_name(cur, "check");
  ```
- after computing the graph, I dump a few elements of the `check` tensor, as well as the sum of the elements:
  https://github.com/ggerganov/ggml/blob/245bdba3c30517ba8c2116803b508fed951c00ed/examples/sam/main.cpp#L696-L712

- dump the same elements in the reference implementation and make sure the numbers are the same or very close
- once I confirm this, I move the `check` tensor further down the graph and repeat the process

**Comment by klosax** [comment]
Similar https://github.com/ggerganov/llama.cpp/issues/2609

**Comment by PABannier** [comment]
@ggerganov Thanks for the answer. Closing the issue.
